README.md
# Hex Describe

This application takes a map generated by *Text Mapper* and a set of
random tables to generate a textual description of the region. It's
ideal if your players are wandering into unprepared regions, and it's
great if you need some seed material to base your work on.

You can provide your own random tables if you have the file online
somewhere in a [Pastebin](https://en.wikipedia.org/wiki/Pastebin) or
shared it from Dropbox, etc.

The format is simple: every word in the map description and every two
word combo from the map description is a potential table in your file.
If it exists, it will be used.

Assuming the following description:

```text
0101 dark-green trees village
```

The description will be generated from any tables that match:

* dark-green
* trees
* village

It would make sense to just provide tables for "trees" and "village",
for example.

Tables looks like this:

```text
;trees
1,some trees
1,you encounter [forest monster]

;forest monster
3,[3d6] bandits
1,an elf
```

A semicolon and some text begin a new table. A number, a comma, and
some text are an entry in he table. The text needs to be on one line.
The numbers are relative probabilities. The chances for an encounter
in the forest are thus 50% and the chances to encounter an elf, if you
are encountering anything at all, are 25%. The example also shows how
you can link from one table to another using square brackets.

Square brackets are also used for dice rolls. A dice roll can look
like this: 3d6, 3d6+5 3d6x10, or 3d6x10+5.

There's an built-in help page with more details for end users. If you
intend to host the application yourself, use `perldoc hex-describe.pl`
to get a more technical documentation based on the comments in the
code.

[Try it](https://campaignwiki.org/hex-describe).

The app comes with a tutorial built in. See the
[Help](https://campaignwiki.org/hex-describe/help) link.

## Dependencies

Perl Modules (or Debian modules):

* Array::Utils or libarray-utils-perl
* IO::Socket::SSL or libio-socket-ssl-perl
* LWP::UserAgent or liblwp-useragent-perl
* List::MoreUtils or liblist-moreutils-perl
* Modern::Perl or libmodern-perl-perl
* Mojolicious or libmojolicious-perl
* Text::Autoformat or libtext-autoformat-perl

The IO::Socket::SSL dependency means that you’ll need OpenSSL
development libraries installed as well: openssl-devel or equivalent,
depending on your package manager.

To install from the working directory (which will also install all the
dependencies) use cpan or cpanm.

Example:

```bash
cpanm .
```

## Installation

Use cpan or cpanm to install Game::HexDescribe.

Using `cpan`:

```shell
cpan Game::HexDescribe
```

Manual install:

```shell
perl Makefile.PL
make
make install
```

## Configuration

In the directory you want to run it from, you may create a config file
named `hex-describe.conf` like the following:

```perl
{
  # error, warn, info, or debug
  loglevel => 'debug',
  # undef means stderr, or a file name
  logfile => undef,
  # undef means the default directory, or a directory name
  contrib => 'share',
  # the URL where you run Text Mapper to generate maps (optional)
  text_mapper_url => 'http://localhost:3010',
  # the URL where you run Face Generator to generate faces (optional)
  face_generator_url => 'http://localhost:3020',
}
```

## Development

As a developer, morbo makes sure to restart the web app whenever a
file changes:

```bash
morbo --mode development --listen "http://*:3000" script/hex-describe
```

Alternatively:

```bash
script/hex-describe daemon --mode development --listen "http://*:3000"
```

## Docker

If you want to experiment with a clean Perl environment in order to
check whether the dependencies are OK, make sure you have Docker
installed, are part of the docker group, and get the latest-perl
image:

```bash
sudo apt install docker.io
sudo adduser $(whoami) docker
# if groups doesn’t show docker, you need to log in again
su - $(whoami)
# this binds the working directory to /app inside the image
docker run -it --rm -v $(pwd):/app perl:latest /bin/bash
# now we’re root inside the image
cd /app
cpanm File::ShareDir::Install
cpanm .
```


Changes
Next

- Add option to not show the face.
- Add < > and = to the arithmetic operators.
- Documented the arithmetic rule. :)
- Fix repeated clicking of the Regenerate button.
- More rules for the Schroeder tables.

1.02

- Fix EOL handling for Windows

1.01

- Add a subcommand for the command line: "hex-describe rule --help"
  for more information.
- More rules for the Schroeder tables.
- Add Forke's tables (German).

1.00

- New, extracted from a previous repository where it was mixed with
  other applications.
  https://alexschroeder.ch/cgit/hex-mapping


lib/Game/HexDescribe/Command/rule.pm
# Copyright (C) 2022  Alex Schroeder <alex@gnu.org>
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

=head1 NAME

Game::HexDescribe::Command::rule

=head1 SYNOPSIS

hex-describe rule [--table=B<table>] [--rule=B<rule>] [--limit=B<number>]

hex-describe rule --help

=head1 DESCRIPTION

Prints the random results of rules.

=head1 OPTIONS

C<--help> prints the man page.

C<--table> specifies the table to load. Without this option, print all the
tables there are.

C<--rule> specifies the rule to use.

C<--text> specifies the text to use (with rules in square brackets).

Without C<--rule> or C<--text>, print all the rules there are.

C<--limit> limits the output to a certain number of entries. The default is 10.

C<--separator> changes the separator between entries. The default is two
newlines, three dashes, and two newlines ("\n\n---\n\n").

=head1 EXAMPLES

Find all orc related rules:

C<hex-describe rule --table=schroeder | grep orc>

Print an orc tribe:

C<hex-describe rule --table=schroeder --rule="orcs" --limit=1>

Print ten orc names:

C<hex-describe rule --table=schroeder --rule="orc name" --separator=\n

=cut

package Game::HexDescribe::Command::rule;

use Modern::Perl '2018';
use Mojo::Base 'Mojolicious::Command';
use Pod::Simple::Text;
use Getopt::Long qw(GetOptionsFromArray);
use Game::HexDescribe::Utils qw(markdown describe_text parse_table load_table list_tables);
use Role::Tiny;
binmode(STDOUT, ':utf8');

has description => 'Print the output of a rule to STDOUT';

has usage => sub { my $self = shift; $self->extract_usage };

sub run {
  my ($self, @args) = @_;
  my ($help, $table, $rule, $text, $limit, $separator);
  GetOptionsFromArray (
    \@args,
    "help" => \$help,
    "table=s" => \$table,
    "rule=s" => \$rule,
    "text=s" => \$text,
    "limit=i" => \$limit,
    "separator=s" => \$separator);
  if ($help) {
    seek(DATA, 0, 0); # read from this file
    my $parser = Pod::Simple::Text->new();
    $parser->output_fh(*STDOUT);
    $parser->parse_lines(<DATA>);
    return 1;
  }
  my $dir = $self->app->config('contrib');
  if (not $table) {
    say for list_tables($dir);
    return 1;
  }
  if (not $rule and not $text) {
    say for keys %{parse_table(load_table($table, $dir))};
    return 1;
  }
  my $data = parse_table(load_table($table, $dir));
  $text //= "[$rule]\n" x ($limit || 10);
  say join("\n", markdown(describe_text($text, $data), $separator));
  1;
}

1;

__DATA__


Makefile.PL
use 5.008004;
use ExtUtils::MakeMaker;
use File::ShareDir::Install;

install_share 'share';

WriteMakefile(
  NAME             => 'Game::HexDescribe',
  VERSION_FROM     => 'lib/Game/HexDescribe.pm',
  ABSTRACT_FROM    => 'lib/Game/HexDescribe.pm',
  AUTHOR           => 'Alex Schroeder',
  LICENSE          => 'agpl_3',
  MIN_PERL_VERSION => '5.26.0', # Modern::Perl '2018'
  EXE_FILES        => [
    'script/hex-describe',
  ],
  PREREQ_PM => {
    'strict' => 0,
    'warnings' => 0,
    'Modern::Perl' => 1.20180701, # for '2018'
    'Mojolicious' => 0,
    'Array::Utils' => 0,
    'List::MoreUtils' => 0,
    'Text::Autoformat' => 0,
    'File::ShareDir' => 0,
    'IO::Socket::SSL' => 0, # if text-mapper requires https
  },
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '6.52',
    'File::ShareDir::Install' => 0,
  },
  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      repository => {
	type => 'git',
	url  => 'https://alexschroeder.ch/cgit/hex-describe',
	web  => 'https://alexschroeder.ch/cgit/hex-describe/about/',
      },
    },
  },
  clean => {
    FILES => "test-[0-9][0-9][0-9][0-9]",
  },
);

package MY;
use File::ShareDir::Install;

sub postamble {
  my $self = shift;
  my @ret = File::ShareDir::Install::postamble($self);
  push @ret, <<'EOT';
check:
	find lib -name '*.pm' -exec perl -Ilib -c '{}' ';' -exec podchecker '{}' ';'

# Assuming you want to run Text Mapper on port 3010 and Face Generator on port
# 3020, start Hex Describe on port 3000.
run:
	morbo --mode development --listen http://*:3000 script/hex-describe
EOT
  return join "\n", @ret;
}
